### DISCUSSION TITLE
OneVsOneClassifier for SLIM doesnt't work

### DISCUSSION THREAD

**Discussion by hermann74** [scikit-mine/scikit-mine#140] [Q&A]
This piece of code doesn't work for 3 classes dataset
```
ovo = OneVsOneClassifier(SLIM())
ovo.fit(D, y=labels)
ovo.decision_function(new_D) # To get  score
ovo.predict(new_D) # to predict class
```
Some called functions :

-  `OneVsOneClassifier.decision_function()`

```python
def decision_function(self, X):
    ...
    predictions = np.vstack([est.predict(Xi) for est, Xi in zip(self.estimators_, Xs)]).T
    confidences = np.vstack([_predict_binary(est, Xi) for est, Xi in zip(self.estimators_, Xs)]).T
    Y = _ovr_decision_function(predictions, confidences, len(self.classes_))
    if self.n_classes_ == 2:
        return Y[:, 1]
    return Y
```

-  `OneVsOneClassifier.predict()`
        
```python
def predict(self, X):
    Y = self.decision_function(X)
    if self.n_classes_ == 2:
        thresh = _threshold_for_binary_predict(self.estimators_[0])
    return self.classes_[(Y > thresh).astype(int)]
    indmax= Y.argmax(axis=1)
    pred = self.classes_[indmax]
    return pred 
```
-  `OneVsOneClassifier._threshold_for_binary_predict()`

```python
def _threshold_for_binary_predict(estimator):
    """Threshold for predictions from binary estimator."""
    if hasattr(estimator, "decision_function") and is_classifier(estimator):
        return 0.0
    else:
        # predict_proba threshold
        return 0.5
```

 

 - For k=2 classes, a threshold is necessary to compute classes from scores 
 
In order to get OvO scores, predictions are  computed from slim instances fitted on Xi dataset (class j VS classe k) thanks to slim.predict(Xi).

An unsuccessful attempt was to add a .predict method to SLIM and modify the output of `slim.decision_function`  to return numpy array
Here these SLIM methods : 
- `SLIM.decision_function()`
```python
def decision_function(self, D):
    prob = _fonc(self.get_code_length(D))
    prob = np.array(prob)  # Convert from pd.Series to np vector
    prob = np.expand_dims(prob, axis=1)
    return prob
```

- `SLIM.predict()`

```python
 def predict(self, X):
     scores = self.decision_function(X)
     indclass= 1 * (scores > 0.25).ravel()
     return self.classes_[indclass]
 ```

**In short, difficult to predict with OvO and even more difficult with 2 classes !**

**Comment by hermann74** [comment]
the example used for these attempts : 

    transactions = [
        ['milk', 'bananas'],
        ['tea', 'New York Times', 'El Pais'],
        ['New York Times'],
        ['El Pais', 'The Economist'],
        ['milk', 'tea'],
        ['croissant', 'tea'],
        ['croissant', 'chocolatine', 'milk'],
    ]

    labels_st = [
        'foodstore',
        'newspaper',
        'newspaper',
        'newspaper',
        'foodstore',
        'bakery',
        'bakery',
    ]

    new_transaction = [
        ['croissant', 'chocolatine', 'milk'],
        ['El Pais', 'tea']
    ]
    new_labels = [
        'bakery',
        'newspaper'
    ]
    
    te = TransactionEncoder()
    D = te.fit(transactions).transform(transactions)
    new_D = te.transform(new_transaction)
    le = LabelEncoder()
    labels = le.fit_transform(labels_st)


     class TransactionEncoder(MultiLabelBinarizer):  # pandas DataFrames are easier to read ;)
            def transform(self, X):
                _X = super().transform(X)
                return pd.DataFrame(data=_X, columns=self.classes_)
